contracts/utils/cryptography/TrieProof.sol

```diff
@@ -0,0 +1,226 @@
+     * malformed input leads to RLP decoding errors.
+     */
+    function tryTraverse(
+        bytes32 root,
+        bytes memory key,
+        bytes[] memory proof
+    ) internal pure returns (bytes memory value, ProofError err) {
+        if (key.length == 0) return (_emptyBytesMemory(), ProofError.EMPTY_KEY);
+
+        // Expand the key
+        bytes memory keyExpanded = key.toNibbles();
+
+        bytes32 currentNodeId;
+        uint256 currentNodeIdLength;
+
+        // Free memory pointer cache
+        Memory.Pointer fmp = Memory.getFreeMemoryPointer();
+
+        // Traverse proof
+        uint256 keyIndex = 0;
+        uint256 proofLength = proof.length;
+        for (uint256 i = 0; i < proofLength; ++i) {
+            // validates the encoded node matches the expected node id
+            bytes memory encoded = proof[i];
+            if (keyIndex == 0) {
+                // Root node must match root hash
+                if (keccak256(encoded) != root) return (_emptyBytesMemory(), ProofError.INVALID_ROOT);
+            } else if (encoded.length >= 32) {
+                // Large nodes are stored as hashes
+                if (currentNodeIdLength != 32 || keccak256(encoded) != currentNodeId)
+                    return (_emptyBytesMemory(), ProofError.INVALID_LARGE_NODE);
+            } else {
+                // Small nodes must match directly
+                if (currentNodeIdLength != encoded.length || bytes32(encoded) != currentNodeId)
+                    return (_emptyBytesMemory(), ProofError.INVALID_SHORT_NODE);
+            }
+
+            // decode the current node as an RLP list, and process it
+            Memory.Slice[] memory decoded = encoded.decodeList();
+            if (decoded.length == BRANCH_NODE_LENGTH) {
+                // If we've consumed the entire key, the value must be in the last slot
+                // Otherwise, continue down the branch specified by the next nibble in the key
+                if (keyIndex == keyExpanded.length) {
+                    return _validateLastItem(decoded[EVM_TREE_RADIX], proofLength, i);
+                } else {
+                    bytes1 branchKey = keyExpanded[keyIndex];
+                    (currentNodeId, currentNodeIdLength) = _getNodeId(decoded[uint8(branchKey)]);
+                    keyIndex += 1;
+                }
+            } else if (decoded.length == LEAF_OR_EXTENSION_NODE_LENGTH) {
+                bytes memory path = decoded[0].readBytes().toNibbles(); // expanded path
+                if (path.length == 0) {
+                    return (_emptyBytesMemory(), ProofError.EMPTY_PATH);
+                }
+                uint8 prefix = uint8(path[0]);
+                Memory.Slice keyRemainder = keyExpanded.asSlice().slice(keyIndex); // Remaining key to match
+                Memory.Slice pathRemainder = path.asSlice().slice(2 - (prefix % 2)); // Path after the prefix
```

Comment:
This `.slice` call will revert if the provided `path`'s length is 1 and the prefix is even. Shall we just check the following in Line 145?

```solidity
                if (path.length < 2) {
                    return (_emptyBytesMemory(), ProofError.EMPTY_PATH);
                }
```

Especially if this function is `try*`, I guess it's implicit that it shouldn't revert. Although we mention that the function may still revert due to invalid RLP encoding, I don't think this would classify as invalid RLP encoding.

Reply:
> I guess it's implicit that it shouldn't revert

We cannot prevent reverts completelly with the current RLP library. Line 132 will revert if the node is not a properly encoded RLP list. This is documented.

That being said, checking that length is at least 2 sounds like a reasonable check